File: backend/forecast_engine.py
"""
Equilibrium — Burnout & Capacity Forecast Engine (Python Edition)
=================================================================
Pure-Python 7-day strain/overload prediction module with zero
third-party dependencies.

Polarity Standard:
    0–100 measures STRAIN/OVERLOAD (higher score = higher stress).
    Clustered deadlines, sleep debt, and social isolation raise strain.
    Recovery blocks and social interaction lower strain.

Usage:
    from forecast_engine import compute_forecast
    forecast = compute_forecast(tasks, sleep_logs, social_events, recovery_blocks)

Standalone test run:
    python forecast_engine.py
"""

from datetime import datetime, timedelta

DEFAULT_CONFIG = {
    "baseline": 30,
    "academic_weight": 16,
    "task_weight": 5,
    "sleep_weight": 6,
    "sleep_debt_carryover": 0.5,
    "target_sleep": 8.0,
    "isolation_weight": 4,
    "isolation_threshold": 2,
    "social_relief": -10,
    "recovery_weight": 10,
    "carryover_weight": 0.3,
}

DAYS_OF_WEEK = ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]


def compute_forecast(tasks=None, sleep_logs=None, social_events=None, recovery_blocks=None, config=None):
    """Calculate 7-day strain trajectory and contributing factors."""
    tasks = tasks or []
    sleep_logs = sleep_logs or []
    social_events = social_events or []
    recovery_blocks = recovery_blocks or []

    cfg = dict(DEFAULT_CONFIG)
    if config:
        cfg.update(config)

    forecast = []
    prev_score = None
    running_sleep_debt = 0.0
    days_since_social = 1

    base_date = cfg.get("base_date")
    if base_date:
        if isinstance(base_date, str):
            today = datetime.fromisoformat(base_date)
        else:
            today = base_date
    else:
        today = datetime.now()

    for i in range(7):
        factors = []
        current_date = today + timedelta(days=i)
        day_label = "Today" if i == 0 else ("Tomorrow" if i == 1 else DAYS_OF_WEEK[current_date.weekday()])

        # 1. Base capacity/strain
        day_score = float(cfg["baseline"])

        # 2. Deadlines and tasks
        day_tasks = [t for t in tasks if t.get("day_offset", t.get("dayOffset", 0)) == i and not t.get("done") and not t.get("hidden")]
        deadlines = [t for t in day_tasks if t.get("is_deadline", t.get("isDeadline", False))]
        high_tasks = [t for t in day_tasks if not t.get("is_deadline", t.get("isDeadline", False)) and t.get("energy") == "high"]

        deadline_impact = len(deadlines) * cfg["academic_weight"]
        if deadline_impact > 0:
            day_score += deadline_impact
            factors.append({
                "name": "Deadlines",
                "impact": deadline_impact,
                "type": "stressor",
                "description": f"{len(deadlines)} deadline(s) clustering (+{deadline_impact} strain)",
            })

        high_task_impact = len(high_tasks) * cfg["task_weight"]
        if high_task_impact > 0:
            day_score += high_task_impact
            factors.append({
                "name": "High Energy Tasks",
                "impact": high_task_impact,
                "type": "stressor",
                "description": f"{len(high_tasks)} heavy task(s) (+{high_task_impact} strain)",
            })

        # 3. Sleep debt & carryover
        sleep_entry = next((s for s in sleep_logs if s.get("day_offset", s.get("dayOffset")) == i), None)
        sleep_hours = sleep_entry.get("hours", cfg["target_sleep"]) if sleep_entry else cfg["target_sleep"]
        target = sleep_entry.get("target_hours", sleep_entry.get("targetHours", cfg["target_sleep"])) if sleep_entry else cfg["target_sleep"]
        current_deficit = max(0.0, target - sleep_hours)
        running_sleep_debt = current_deficit + running_sleep_debt * cfg["sleep_debt_carryover"]

        if running_sleep_debt >= 0.5:
            sleep_impact = round(running_sleep_debt * cfg["sleep_weight"])
            day_score += sleep_impact
            factors.append({
                "name": "Sleep Debt",
                "impact": sleep_impact,
                "type": "stressor",
                "description": f"{running_sleep_debt:.1f}h cumulative sleep deficit (+{sleep_impact} strain)",
            })
        elif sleep_hours >= target + 0.5:
            sleep_bonus = -6
            day_score += sleep_bonus
            factors.append({
                "name": "Restorative Sleep",
                "impact": sleep_bonus,
                "type": "relief",
                "description": f"{sleep_hours}h well-rested sleep ({sleep_bonus} strain)",
            })

        # 4. Social events & isolation
        day_social = [s for s in social_events if s.get("day_offset", s.get("dayOffset")) == i]
        if day_social:
            days_since_social = 0
            day_score += cfg["social_relief"]
            titles = ", ".join(s.get("title", "Social") for s in day_social)
            factors.append({
                "name": "Social Relief",
                "impact": cfg["social_relief"],
                "type": "relief",
                "description": f"{titles} ({cfg['social_relief']} strain)",
            })
        else:
            days_since_social += 1
            if days_since_social > cfg["isolation_threshold"]:
                iso_impact = (days_since_social - cfg["isolation_threshold"]) * cfg["isolation_weight"]
                day_score += iso_impact
                factors.append({
                    "name": "Social Isolation",
                    "impact": iso_impact,
                    "type": "stressor",
                    "description": f"{days_since_social} days without social contact (+{iso_impact} strain)",
                })

        # 5. Recovery blocks
        day_recovery = [r for r in recovery_blocks if r.get("day_offset", r.get("dayOffset")) == i]
        if day_recovery:
            rec_hours = sum(r.get("duration_hours", r.get("durationHours", 1.0)) for r in day_recovery)
            rec_impact = -round(rec_hours * cfg["recovery_weight"])
            day_score += rec_impact
            titles = ", ".join(r.get("title", "Recovery") for r in day_recovery)
            factors.append({
                "name": "Recovery Time",
                "impact": rec_impact,
                "type": "relief",
                "description": f"{titles} ({rec_impact} strain)",
            })

        # 6. Fatigue momentum carryover
        if prev_score is not None:
            diff = prev_score - cfg["baseline"]
            carryover = round(diff * cfg["carryover_weight"])
            if carryover != 0:
                day_score += carryover
                factors.append({
                    "name": "Fatigue Momentum" if carryover > 0 else "Rest Momentum",
                    "impact": carryover,
                    "type": "stressor" if carryover > 0 else "relief",
                    "description": f"{'+' if carryover > 0 else ''}{carryover} carryover from previous day",
                })

        final_score = max(0, min(100, round(day_score)))
        prev_score = final_score

        if final_score >= 80:
            status = "CRITICAL_OVERLOAD"
        elif final_score >= 60:
            status = "HEAVY_STRAIN"
        else:
            status = "BALANCED"

        forecast.append({
            "day_index": i,
            "day_label": day_label,
            "date_str": current_date.strftime("%b %d"),
            "score": final_score,
            "status_level": status,
            "factors": factors,
        })

    return forecast


if __name__ == "__main__":
    sample_tasks = [
        {"text": "Econ Problem Set", "is_deadline": True, "day_offset": 0},
        {"text": "Midterm Exam", "is_deadline": True, "day_offset": 2},
        {"text": "Term Paper", "is_deadline": True, "day_offset": 3},
    ]
    sample_sleep = [
        {"day_offset": 0, "hours": 7.0},
        {"day_offset": 2, "hours": 4.8},
        {"day_offset": 3, "hours": 4.5},
    ]
    sample_social = [{"day_offset": 0, "title": "Lunch"}, {"day_offset": 4, "title": "Dinner"}]
    sample_recovery = [{"day_offset": 5, "title": "Beach walk", "duration_hours": 2.0}]

    results = compute_forecast(sample_tasks, sample_sleep, sample_social, sample_recovery)
    print("--- 7-Day Forecast Sample Output ---")
    for day in results:
        print(f"{day['day_label']:<10} | Score: {day['score']:<3} | Status: {day['status_level']:<18} | Factors: {len(day['factors'])}")

File: _deprecated/forecastEngine.js
/**
 * Equilibrium — Burnout & Capacity Forecast Engine (Backend Edition)
 * =================================================================
 * Pure, deterministic forecasting utility calculating a 7-day
 * strain/overload prediction (0–100 score).
 *
 * Polarity Standard:
 *   0–100 measures STRAIN/OVERLOAD.
 *   Higher score = higher stress / risk of burnout.
 *   Clustered deadlines, sleep deficits, and isolation spike the curve upward.
 *   Recovery blocks and social connection pull the curve downward.
 *
 * Can be run standalone:
 *   node backend/forecastEngine.js
 */

export const DEFAULT_CONFIG = {
  baseline: 30,           // Normal baseline strain
  academicWeight: 16,     // Strain per academic deadline
  taskWeight: 5,          // Strain per high-energy regular task
  sleepWeight: 6,         // Strain per hour of sleep deficit
  sleepDebtCarryover: 0.5,// Half of previous night's debt carries over
  targetSleep: 8.0,       // Target restorative sleep
  isolationWeight: 4,     // Strain penalty per day past isolation threshold
  isolationThreshold: 2,  // Grace period before isolation penalty begins
  socialRelief: -10,      // Relief points from a social event
  recoveryWeight: 10,     // Relief points per recovery block
  carryoverWeight: 0.3,   // 30% carryover of previous day's strain delta
};

/**
 * Computes a 7-day burnout/strain forecast.
 *
 * @param {Array} tasks List of task objects with { dayOffset, isDeadline, energy, done, hidden }
 * @param {Array} sleepLogs List of { dayOffset, hours, targetHours }
 * @param {Array} socialEvents List of { dayOffset, title, durationHours }
 * @param {Array} recoveryBlocks List of { dayOffset, title, durationHours }
 * @param {Object} config Override weights and thresholds (accepts baseDate for deterministic tests)
 * @returns {Array} 7-day forecast array [{ dayIndex, dayLabel, dateStr, score, statusLevel, factors }]
 */
export function computeForecast(
  tasks = [],
  sleepLogs = [],
  socialEvents = [],
  recoveryBlocks = [],
  config = {}
) {
  const cfg = { ...DEFAULT_CONFIG, ...config };
  const forecast = [];
  let previousDayScore = null;
  let runningSleepDebt = 0;
  let daysSinceSocial = 1; // Default to 1 day since last contact

  const daysOfWeek = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
  const today = cfg.baseDate ? new Date(cfg.baseDate) : new Date();

  for (let i = 0; i < 7; i++) {
    const factors = [];
    const dateObj = new Date(today);
    dateObj.setDate(today.getDate() + i);
    const dayLabel = i === 0 ? 'Today' : i === 1 ? 'Tomorrow' : daysOfWeek[dateObj.getDay()];

    // 1. Base capacity/strain
    let dayScore = cfg.baseline;

    // 2. Academic Deadlines & Work Tasks
    const dayTasks = tasks.filter((t) => (t.dayOffset ?? 0) === i && !t.done && !t.hidden);
    const deadlines = dayTasks.filter((t) => t.isDeadline);
    const highTasks = dayTasks.filter((t) => !t.isDeadline && t.energy === 'high');

    const deadlineImpact = deadlines.length * cfg.academicWeight;
    if (deadlineImpact > 0) {
      dayScore += deadlineImpact;
      factors.push({
        name: 'Deadlines',
        impact: +deadlineImpact,
        type: 'stressor',
        description: `${deadlines.length} deadline${deadlines.length > 1 ? 's' : ''} clustering (+${deadlineImpact} strain)`,
      });
    }

    const highTaskImpact = highTasks.length * cfg.taskWeight;
    if (highTaskImpact > 0) {
      dayScore += highTaskImpact;
      factors.push({
        name: 'High Energy Tasks',
        impact: +highTaskImpact,
        type: 'stressor',
        description: `${highTasks.length} heavy task${highTasks.length > 1 ? 's' : ''} (+${highTaskImpact} strain)`,
      });
    }

    // 3. Sleep Debt & Carryover
    const sleepEntry = sleepLogs.find((s) => s.dayOffset === i);
    const sleepHours = sleepEntry ? sleepEntry.hours : cfg.targetSleep;
    const target = (sleepEntry && sleepEntry.targetHours) || cfg.targetSleep;
    const currentNightDeficit = Math.max(0, target - sleepHours);

    // Accumulate sleep debt with dampening carryover
    runningSleepDebt = currentNightDeficit + runningSleepDebt * cfg.sleepDebtCarryover;

    if (runningSleepDebt >= 0.5) {
      const sleepImpact = Math.round(runningSleepDebt * cfg.sleepWeight);
      dayScore += sleepImpact;
      factors.push({
        name: 'Sleep Debt',
        impact: +sleepImpact,
        type: 'stressor',
        description: `${runningSleepDebt.toFixed(1)}h cumulative sleep deficit (+${sleepImpact} strain)`,
      });
    } else if (sleepHours >= target + 0.5) {
      const sleepBonus = -6;
      dayScore += sleepBonus;
      factors.push({
        name: 'Restorative Sleep',
        impact: sleepBonus,
        type: 'relief',
        description: `${sleepHours}h well-rested sleep (${sleepBonus} strain)`,
      });
    }

    // 4. Social Interaction & Isolation Penalty
    const daySocial = socialEvents.filter((s) => s.dayOffset === i);
    if (daySocial.length > 0) {
      daysSinceSocial = 0;
      dayScore += cfg.socialRelief;
      factors.push({
        name: 'Social Relief',
        impact: cfg.socialRelief,
        type: 'relief',
        description: `${daySocial.map((s) => s.title).join(', ')} (${cfg.socialRelief} strain)`,
      });
    } else {
      daysSinceSocial += 1;
      if (daysSinceSocial > cfg.isolationThreshold) {
        const isoImpact = (daysSinceSocial - cfg.isolationThreshold) * cfg.isolationWeight;
        dayScore += isoImpact;
        factors.push({
          name: 'Social Isolation',
          impact: +isoImpact,
          type: 'stressor',
          description: `${daysSinceSocial} days without social contact (+${isoImpact} strain)`,
        });
      }
    }

    // 5. Recovery Blocks
    const dayRecovery = recoveryBlocks.filter((r) => r.dayOffset === i);
    if (dayRecovery.length > 0) {
      const recTotalHours = dayRecovery.reduce((acc, r) => acc + (r.durationHours || 1), 0);
      const recImpact = -Math.round(recTotalHours * cfg.recoveryWeight);
      dayScore += recImpact;
      factors.push({
        name: 'Recovery Time',
        impact: recImpact,
        type: 'relief',
        description: `${dayRecovery.map((r) => r.title).join(', ')} (${recImpact} strain)`,
      });
    }

    // 6. Carryover from Previous Day's Strain
    if (previousDayScore !== null) {
      const diffFromBaseline = previousDayScore - cfg.baseline;
      const carryover = Math.round(diffFromBaseline * cfg.carryoverWeight);
      if (carryover !== 0) {
        dayScore += carryover;
        factors.push({
          name: carryover > 0 ? 'Fatigue Momentum' : 'Rest Momentum',
          impact: carryover,
          type: carryover > 0 ? 'stressor' : 'relief',
          description: `${carryover > 0 ? '+' : ''}${carryover} carryover from previous day`,
        });
      }
    }

    // Clamp score strictly between 0 and 100
    const finalScore = Math.max(0, Math.min(100, Math.round(dayScore)));
    previousDayScore = finalScore;

    let statusLevel = 'BALANCED';
    if (finalScore >= 80) statusLevel = 'CRITICAL_OVERLOAD';
    else if (finalScore >= 60) statusLevel = 'HEAVY_STRAIN';

    forecast.push({
      dayIndex: i,
      dayLabel,
      dateStr: dateObj.toLocaleDateString('en-US', { month: 'short', day: 'numeric' }),
      score: finalScore,
      statusLevel,
      factors,
    });
  }

  return forecast;
}

/**
 * Self-contained verification suite running the 4 required test cases:
 * a) High sleep debt accumulation
 * b) Zero social contact penalty
 * c) Clustered deadlines spike
 * d) Carryover dampening across consecutive days
 */
export function runVerificationTests() {
  const results = [];

  function assert(testName, condition, detail = '') {
    results.push({ testName, passed: Boolean(condition), detail });
    if (!condition) {
      console.error(`❌ FAIL: ${testName} - ${detail}`);
    } else {
      console.log(`✅ PASS: ${testName}`);
    }
  }

  // ─────────────────────────────────────────────────────────────
  // Test A: High sleep debt accumulation
  // ─────────────────────────────────────────────────────────────
  {
    const chronicSleepDeprivation = [
      { dayOffset: 0, hours: 4.5, targetHours: 8 },
      { dayOffset: 1, hours: 4.5, targetHours: 8 },
      { dayOffset: 2, hours: 4.5, targetHours: 8 },
      { dayOffset: 3, hours: 4.5, targetHours: 8 },
      { dayOffset: 4, hours: 4.5, targetHours: 8 },
      { dayOffset: 5, hours: 4.5, targetHours: 8 },
      { dayOffset: 6, hours: 4.5, targetHours: 8 },
    ];
    const forecast = computeForecast([], chronicSleepDeprivation, [{ dayOffset: 0, title: 'Checkin' }]);
    const day0Score = forecast[0].score;
    const day2Score = forecast[2].score;
    const day4Score = forecast[4].score;

    assert(
      'Test A: High sleep debt accumulation increases strain over time',
      day4Score > day2Score && day2Score > day0Score,
      `Day 0: ${day0Score}, Day 2: ${day2Score}, Day 4: ${day4Score}`
    );
  }

  // ─────────────────────────────────────────────────────────────
  // Test B: Zero social contact penalty
  // ─────────────────────────────────────────────────────────────
  {
    const withSocial = computeForecast(
      [],
      [],
      [
        { dayOffset: 1, title: 'Dinner with friends' },
        { dayOffset: 3, title: 'Coffee hangout' },
        { dayOffset: 5, title: 'Game night' },
      ]
    );

    const isolated = computeForecast([], [], []); // Zero social contact all week

    const day5WithSocial = withSocial[5].score;
    const day5Isolated = isolated[5].score;
    const isolationFactor = isolated[5].factors.find((f) => f.name === 'Social Isolation');

    assert(
      'Test B: Zero social contact triggers isolation penalty raising strain',
      day5Isolated > day5WithSocial && Boolean(isolationFactor),
      `Isolated day 5 score: ${day5Isolated} vs Connected day 5: ${day5WithSocial} (factor: ${isolationFactor?.impact})`
    );
  }

  // ─────────────────────────────────────────────────────────────
  // Test C: Clustered deadlines spike
  // ─────────────────────────────────────────────────────────────
  {
    const tasksWithCrunch = [
      { id: 1, dayOffset: 2, isDeadline: true, energy: 'high', done: false },
      { id: 2, dayOffset: 2, isDeadline: true, energy: 'high', done: false },
      { id: 3, dayOffset: 2, isDeadline: true, energy: 'high', done: false },
    ];

    const forecast = computeForecast(tasksWithCrunch, [], [{ dayOffset: 2, title: 'Sync' }]);
    const calmDayScore = forecast[0].score;
    const crunchDayScore = forecast[2].score;

    assert(
      'Test C: Clustered deadlines cause dramatic strain spike (>= 30 pts)',
      crunchDayScore - calmDayScore >= 30,
      `Calm day: ${calmDayScore}, Crunch day: ${crunchDayScore} (Spike: +${crunchDayScore - calmDayScore})`
    );
  }

  // ─────────────────────────────────────────────────────────────
  // Test D: Carryover dampening across consecutive days
  // ─────────────────────────────────────────────────────────────
  {
    // A huge crunch on Day 0, then zero tasks for days 1, 2, 3
    const tasks = [
      { id: 1, dayOffset: 0, isDeadline: true, energy: 'high', done: false },
      { id: 2, dayOffset: 0, isDeadline: true, energy: 'high', done: false },
      { id: 3, dayOffset: 0, isDeadline: true, energy: 'high', done: false },
      { id: 4, dayOffset: 0, isDeadline: true, energy: 'high', done: false },
    ];

    // Maintain regular social sync so isolation does not confound carryover dampening
    const social = [0, 1, 2, 3, 4, 5, 6].map((d) => ({ dayOffset: d, title: 'Chat' }));

    const forecast = computeForecast(tasks, [], social);
    const day0Score = forecast[0].score; // Very high
    const day1Score = forecast[1].score;
    const day2Score = forecast[2].score;
    const day3Score = forecast[3].score;

    const carryoverDay1 = forecast[1].factors.find((f) => f.name === 'Fatigue Momentum')?.impact ?? 0;
    const carryoverDay2 = forecast[2].factors.find((f) => f.name === 'Fatigue Momentum')?.impact ?? 0;
    const carryoverDay3 = forecast[3].factors.find((f) => f.name === 'Fatigue Momentum')?.impact ?? 0;

    assert(
      'Test D: Carryover dampens monotonically across consecutive days after crunch',
      day0Score > day1Score && day1Score > day2Score && day2Score > day3Score && carryoverDay1 > carryoverDay2 && carryoverDay2 >= carryoverDay3,
      `Scores: D0=${day0Score}, D1=${day1Score}, D2=${day2Score}, D3=${day3Score}; Carryover: D1=+${carryoverDay1}, D2=+${carryoverDay2}, D3=+${carryoverDay3}`
    );
  }

  return results;
}

// If executed directly in node: `node backend/forecastEngine.js`
const nodeProcess = globalThis?.process;
if (nodeProcess?.argv && nodeProcess.argv[1]?.includes('forecastEngine.js')) {
  console.log('--- Running Backend Forecast Engine Verification Tests ---');
  const allResults = runVerificationTests();
  const allPassed = allResults.every((r) => r.passed);
  console.log(`--- Result: ${allPassed ? 'ALL TESTS PASSED' : 'SOME TESTS FAILED'} (${allResults.length} tests) ---`);
  if (!allPassed) nodeProcess.exit(1);
}
